File: backend/app/events/event_catalog.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

EVENT_CLASSIFICATIONS = ("BIOSECURITY", "ANIMAL", "VEHICLE", "SYSTEM")
EVENT_SEVERITIES = ("LOW", "MEDIUM", "HIGH", "CRITICAL")

EVENT_TYPE_ALIASES = {
    "PROCESS_VIOLATION": "BIOSECURITY_PROCESS_VIOLATION",
}


@dataclass(frozen=True)
class EventExplanation:
    title: str
    description: str
    recommended_action: str

    def to_dict(self) -> dict[str, str]:
        return {
            "title": self.title,
            "description": self.description,
            "recommendedAction": self.recommended_action,
        }


@dataclass(frozen=True)
class EventCatalogEntry:
    classification: str
    severity: str
    explanation: EventExplanation


EVENT_CATALOG: dict[str, EventCatalogEntry] = {
    "UNIFORM_VIOLATION": EventCatalogEntry(
        classification="BIOSECURITY",
        severity="MEDIUM",
        explanation=EventExplanation(
            title="Sai đồng phục vùng",
            description="Người đi vào vùng sạch nhưng không sử dụng đồng phục phù hợp.",
            recommended_action="Kiểm tra quy trình thay đồ và sát trùng.",
        ),
    ),
    "NO_HAND_SANITIZATION": EventCatalogEntry(
        classification="BIOSECURITY",
        severity="HIGH",
        explanation=EventExplanation(
            title="Không sát trùng tay",
            description="Người bỏ qua bước rửa tay sát trùng trước khi vào vùng kiểm soát.",
            recommended_action="Nhắc nhở và kiểm tra lại quy trình sát trùng tay tại điểm giám sát.",
        ),
    ),
    "NO_BOOT_SANITIZATION": EventCatalogEntry(
        classification="BIOSECURITY",
        severity="HIGH",
        explanation=EventExplanation(
            title="Không sát trùng ủng",
            description="Người bỏ qua bước sát trùng ủng trước khi vào vùng sạch.",
            recommended_action="Kiểm tra khay sát trùng ủng và giám sát tuân thủ quy trình.",
        ),
    ),
    "ZONE_INTRUSION": EventCatalogEntry(
        classification="BIOSECURITY",
        severity="HIGH",
        explanation=EventExplanation(
            title="Xâm nhập vùng cấm",
            description="Người hoặc đối tượng đi vào vùng bị hạn chế hoặc cấm tuyệt đối.",
            recommended_action="Xác minh danh tính, ghi nhận vi phạm và áp dụng biện pháp cách ly.",
        ),
    ),
    "BIOSECURITY_PROCESS_VIOLATION": EventCatalogEntry(
        classification="BIOSECURITY",
        severity="CRITICAL",
        explanation=EventExplanation(
            title="Vi phạm quy trình an toàn sinh học",
            description="Người bỏ qua một hoặc nhiều bước bắt buộc trong quy trình vào vùng sạch.",
            recommended_action="Dừng di chuyển, đưa người vi phạm quay lại thực hiện đầy đủ quy trình.",
        ),
    ),
    "ANIMAL_INTRUSION": EventCatalogEntry(
        classification="ANIMAL",
        severity="HIGH",
        explanation=EventExplanation(
            title="Động vật xâm nhập",
            description="Động vật không được phép xuất hiện trong vùng sản xuất hoặc vùng sạch.",
            recommended_action="Loại bỏ động vật khỏi vùng và kiểm tra rào chắn, cửa ra vào.",
        ),
    ),
    "VEHICLE_INTRUSION": EventCatalogEntry(
        classification="VEHICLE",
        severity="HIGH",
        explanation=EventExplanation(
            title="Xe vi phạm sát trùng",
            description="Phương tiện vào vùng kiểm soát mà chưa hoàn thành quy trình sát trùng.",
            recommended_action="Dừng xe, yêu cầu sát trùng lại và ghi nhận biển số.",
        ),
    ),
    "CAMERA_OFFLINE": EventCatalogEntry(
        classification="SYSTEM",
        severity="MEDIUM",
        explanation=EventExplanation(
            title="Camera mất kết nối",
            description="Camera không gửi tín hiệu hoặc heartbeat trong thời gian quy định.",
            recommended_action="Kiểm tra nguồn điện, mạng và thiết bị tại vị trí camera.",
        ),
    ),
}

CATEGORY_CLASSIFICATION_MAP = {
    "compliance_violation": "BIOSECURITY",
    "animal_intrusion": "ANIMAL",
    "atsh_violation": "BIOSECURITY",
    "workflow_violation": "BIOSECURITY",
    "camera_offline": "SYSTEM",
    "rule_engine": "BIOSECURITY",
}


def normalize_event_type(event_type: str | None) -> str | None:
    if not event_type:
        return None
    normalized = str(event_type).strip().upper()
    return EVENT_TYPE_ALIASES.get(normalized, normalized)


def resolve_event_classification(
    event_type: str | None,
    *,
    category: str | None = None,
) -> str:
    normalized = normalize_event_type(event_type)
    if normalized and normalized in EVENT_CATALOG:
        return EVENT_CATALOG[normalized].classification
    if category and category in CATEGORY_CLASSIFICATION_MAP:
        return CATEGORY_CLASSIFICATION_MAP[category]
    return "SYSTEM"


def resolve_event_severity(
    event_type: str | None,
    *,
    fallback: str | None = None,
) -> str:
    normalized = normalize_event_type(event_type)
    if normalized and normalized in EVENT_CATALOG:
        return EVENT_CATALOG[normalized].severity
    if fallback:
        upper = str(fallback).upper()
        if upper in EVENT_SEVERITIES:
            return upper
        legacy_map = {
            "CRITICAL": "CRITICAL",
            "DANGER": "CRITICAL",
            "HIGH": "HIGH",
            "WARNING": "MEDIUM",
            "MEDIUM": "MEDIUM",
            "LOW": "LOW",
            "INFO": "LOW",
        }
        return legacy_map.get(upper, "MEDIUM")
    return "MEDIUM"


def build_event_explanation(
    event_type: str | None,
    *,
    rule_name: str | None = None,
    zone_name: str | None = None,
) -> dict[str, str]:
    normalized = normalize_event_type(event_type)
    entry = EVENT_CATALOG.get(normalized or "")
    if entry:
        explanation = entry.explanation.to_dict()
        if rule_name and normalized == "BIOSECURITY_PROCESS_VIOLATION":
            explanation["title"] = rule_name
        if zone_name:
            explanation["description"] = f"{explanation['description']} Vùng: {zone_name}."
        return explanation

    title = rule_name or normalized or "Vi phạm tuân thủ"
    return {
        "title": title,
        "description": f"Sự kiện {title} được ghi nhận bởi hệ thống AMS.",
        "recommendedAction": "Kiểm tra camera, xác minh vi phạm và xử lý theo quy trình nội bộ.",
    }


def enrich_event_fields(
    *,
    event_type: str | None,
    category: str | None = None,
    severity: str | None = None,
    rule_name: str | None = None,
    zone_name: str | None = None,
) -> dict[str, Any]:
    normalized_type = normalize_event_type(event_type)
    resolved_severity = resolve_event_severity(normalized_type, fallback=severity)
    classification = resolve_event_classification(normalized_type, category=category)
    explanation = build_event_explanation(
        normalized_type,
        rule_name=rule_name,
        zone_name=zone_name,
    )
    return {
        "event_type": normalized_type or event_type,
        "classification": classification,
        "severity": resolved_severity,
        "title": explanation["title"],
        "description": explanation["description"],
        "recommendedAction": explanation["recommendedAction"],
        "explanation": explanation,
    }

File: backend/src/modules/events/eventCatalog.js
export const EVENT_CLASSIFICATIONS = {
  BIOSECURITY: 'BIOSECURITY',
  ANIMAL: 'ANIMAL',
  VEHICLE: 'VEHICLE',
  SYSTEM: 'SYSTEM',
}

export const EVENT_SEVERITIES = {
  LOW: 'LOW',
  MEDIUM: 'MEDIUM',
  HIGH: 'HIGH',
  CRITICAL: 'CRITICAL',
}

export const EVENT_TYPE_ALIASES = {
  PROCESS_VIOLATION: 'BIOSECURITY_PROCESS_VIOLATION',
}

export const EVENT_CATALOG = {
  UNIFORM_VIOLATION: {
    classification: 'BIOSECURITY',
    severity: 'MEDIUM',
    explanation: {
      title: 'Sai đồng phục vùng',
      description: 'Người đi vào vùng sạch nhưng không sử dụng đồng phục phù hợp.',
      recommendedAction: 'Kiểm tra quy trình thay đồ và sát trùng.',
    },
  },
  NO_HAND_SANITIZATION: {
    classification: 'BIOSECURITY',
    severity: 'HIGH',
    explanation: {
      title: 'Không sát trùng tay',
      description: 'Người bỏ qua bước rửa tay sát trùng trước khi vào vùng kiểm soát.',
      recommendedAction: 'Nhắc nhở và kiểm tra lại quy trình sát trùng tay tại điểm giám sát.',
    },
  },
  NO_BOOT_SANITIZATION: {
    classification: 'BIOSECURITY',
    severity: 'HIGH',
    explanation: {
      title: 'Không sát trùng ủng',
      description: 'Người bỏ qua bước sát trùng ủng trước khi vào vùng sạch.',
      recommendedAction: 'Kiểm tra khay sát trùng ủng và giám sát tuân thủ quy trình.',
    },
  },
  ZONE_INTRUSION: {
    classification: 'BIOSECURITY',
    severity: 'HIGH',
    explanation: {
      title: 'Xâm nhập vùng cấm',
      description: 'Người hoặc đối tượng đi vào vùng bị hạn chế hoặc cấm tuyệt đối.',
      recommendedAction: 'Xác minh danh tính, ghi nhận vi phạm và áp dụng biện pháp cách ly.',
    },
  },
  BIOSECURITY_PROCESS_VIOLATION: {
    classification: 'BIOSECURITY',
    severity: 'CRITICAL',
    explanation: {
      title: 'Vi phạm quy trình an toàn sinh học',
      description: 'Người bỏ qua một hoặc nhiều bước bắt buộc trong quy trình vào vùng sạch.',
      recommendedAction: 'Dừng di chuyển, đưa người vi phạm quay lại thực hiện đầy đủ quy trình.',
    },
  },
  ANIMAL_INTRUSION: {
    classification: 'ANIMAL',
    severity: 'HIGH',
    explanation: {
      title: 'Động vật xâm nhập',
      description: 'Động vật không được phép xuất hiện trong vùng sản xuất hoặc vùng sạch.',
      recommendedAction: 'Loại bỏ động vật khỏi vùng và kiểm tra rào chắn, cửa ra vào.',
    },
  },
  VEHICLE_INTRUSION: {
    classification: 'VEHICLE',
    severity: 'HIGH',
    explanation: {
      title: 'Xe vi phạm sát trùng',
      description: 'Phương tiện vào vùng kiểm soát mà chưa hoàn thành quy trình sát trùng.',
      recommendedAction: 'Dừng xe, yêu cầu sát trùng lại và ghi nhận biển số.',
    },
  },
  CAMERA_OFFLINE: {
    classification: 'SYSTEM',
    severity: 'MEDIUM',
    explanation: {
      title: 'Camera mất kết nối',
      description: 'Camera không gửi tín hiệu hoặc heartbeat trong thời gian quy định.',
      recommendedAction: 'Kiểm tra nguồn điện, mạng và thiết bị tại vị trí camera.',
    },
  },
}

export const CLASSIFICATION_LABELS = {
  BIOSECURITY: 'An toàn sinh học',
  ANIMAL: 'Động vật',
  VEHICLE: 'Phương tiện',
  SYSTEM: 'Hệ thống',
}

export const SEVERITY_LABELS = {
  LOW: 'Thông tin',
  MEDIUM: 'Cảnh báo',
  HIGH: 'Mức cao',
  CRITICAL: 'Nghiêm trọng',
}

export function normalizeEventType(eventType) {
  if (!eventType) return null
  const normalized = String(eventType).trim().toUpperCase()
  return EVENT_TYPE_ALIASES[normalized] || normalized
}

export function resolveEventClassification(eventType, category) {
  const normalized = normalizeEventType(eventType)
  if (normalized && EVENT_CATALOG[normalized]) {
    return EVENT_CATALOG[normalized].classification
  }
  if (category === 'animal_intrusion') return 'ANIMAL'
  if (category === 'camera_offline') return 'SYSTEM'
  return 'BIOSECURITY'
}

export function resolveEventSeverity(eventType, fallback) {
  const normalized = normalizeEventType(eventType)
  if (normalized && EVENT_CATALOG[normalized]) {
    return EVENT_CATALOG[normalized].severity
  }
  if (fallback) {
    const upper = String(fallback).toUpperCase()
    if (SEVERITY_LABELS[upper]) return upper
  }
  return 'MEDIUM'
}

export function buildEventExplanation(eventType, { ruleName, zoneName } = {}) {
  const normalized = normalizeEventType(eventType)
  const entry = EVENT_CATALOG[normalized]
  if (!entry) {
    return {
      title: ruleName || normalized || 'Vi phạm tuân thủ',
      description: `Sự kiện ${ruleName || normalized || 'vi phạm'} được ghi nhận bởi hệ thống AMS.`,
      recommendedAction: 'Kiểm tra camera, xác minh vi phạm và xử lý theo quy trình nội bộ.',
    }
  }

  const explanation = { ...entry.explanation }
  if (ruleName && normalized === 'BIOSECURITY_PROCESS_VIOLATION') {
    explanation.title = ruleName
  }
  if (zoneName) {
    explanation.description = `${explanation.description} Vùng: ${zoneName}.`
  }
  return explanation
}

export function enrichEventFields(raw = {}) {
  const eventType = raw.eventType || raw.event_type
  const normalized = normalizeEventType(eventType)
  const severity = resolveEventSeverity(normalized, raw.severity)
  const classification = resolveEventClassification(normalized, raw.category)
  const explanation = buildEventExplanation(normalized, {
    ruleName: raw.ruleName || raw.rule_name,
    zoneName: raw.zoneName || raw.zone_name || raw.zone,
  })

  return {
    eventType: normalized || eventType,
    classification,
    severity,
    severityLabel: SEVERITY_LABELS[severity] || severity,
    classificationLabel: CLASSIFICATION_LABELS[classification] || classification,
    title: raw.title || explanation.title,
    description: raw.description || explanation.description,
    recommendedAction: raw.recommendedAction || explanation.recommendedAction,
    explanation: raw.explanation || explanation,
  }
}
